### PR TITLE
Add Renesas RX72N Envision Kit RTOS Demo using Renesas RXv3 port layer

### PR DESCRIPTION
**Pull request by NoMaY-jp**
Description
-----------
As I told in the pull request https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/104,  I'd like to post a pull request to add Renesas RX72N Envision Kit RTOS Demo using Renesas RXv3 port layer. Build environment, download pages, screen snap shots, etc are written/uploaded in the following page.

[https://github.com/NoMaY-jp/FreeRTOS/wiki/RX72N-Envision-Kit-RTOS-Demo-Programs](https://github.com/NoMaY-jp/FreeRTOS/wiki/RX72N-Envision-Kit-RTOS-Demo-Programs)

Almost code are generated code by Renesas RX SmartConfigurator so that folder structure is different from other Renesas RX RTOS Demo projects. Moreover, in case of FreeRTOS project which is generated by e2 studio, generated FreeRTOS project does not have main() so that the following definition is placed in src/FreeRTOS_Demo/demo_main.h
```
/* Set mainCREATE_SIMPLE_BLINKY_DEMO_ONLY to one to run the simple blinky demo,
or 0 to run the more comprehensive test and demo application. */
#define mainCREATE_SIMPLE_BLINKY_DEMO_ONLY  0
```
Test Steps
-----------
(A) Simple Blinky Demo and Full Demo are checked using using on-board LED with the following 5 cases.

e2 studio + GNURX, e2 studio + CC-RX, e2 studio + ICCRX, CS+ for CC (CC-RX), EWRX (ICCRX)

(B) Full Demo's console input/output are checked using terminal software with the same following 5 cases.

e2 studio + GNURX, e2 studio + CC-RX, e2 studio + ICCRX, CS+ for CC (CC-RX), EWRX (ICCRX)

Both (A) and (B) use the following setting.

($-1) Not define USE_FULL_REGISTER_INITIALISATION
($-2) Define configUSE_TASK_DPFPU_SUPPORT = 1

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by RichardBarry** [comment]
Thanks.  A note for other reviewers - I have already procured the RX72N hardware.

**Comment by NoMaY-jp** [comment]
I'm sorry that I replaced one of source files which are workarounds for the Renesas' code generator 'RX SmartConfigurator' to be robust for future updates of the tool.

**Comment by NoMaY-jp** [comment]
The src/ folder structure is as following.
```
src/
    FreeRTOS_Demo/
        Blinky_Demo/    : all files are RTOSDemo specific.
        Full_Demo/      : all files are RTOSDemo specific.
        ./              : all files are RTOSDemo specific.
    frtos_config/       : generated source files, one file has RTOSDemo specific modofication.
    frtos_skeleton/     : generated source files, no modification.
    frtos_startup/      : generated source files, one file has RTOSDemo specific modofication.
    smc_gen/
        general/        : generated source files, one file has RTOSDemo specific modofication.
        r_bsp/          : generated source files, no modification, but pdf document files are removed.
        r_byteq/        : generated source files, no modification, but pdf document files are removed.
        r_config/       : generated source files, one file has general purpos modification.
        r_dtc_rx/       : generated source files, no modification, but pdf document files are removed.
        r_gpio_rx/      : generated source files, no modification, but pdf document files are removed.
        r_pincfg/       : generated source files, no modification.
        r_sci_rx/       : generated source files, no modification, but pdf document files are removed.
    smc_workaround/
        r_bsp_patch/    : workaround source files, no RTOSDemo specific code.
        r_dtc_rx_patch/ : workaround source files, no RTOSDemo specific code.
        ./              : workaround source files, no RTOSDemo specific code.
    ./                  : linker script file for GNURX, but only GNURX project has linker script file.
```
RTOSDemo specific files and generated files which have RTOSDemo specific modofication are as followings.
```
src/FreeRTOS_Demo/Blinky_Demo/main_blinky.c : derived from RX700_RX71M_RSK_GCC_e2studio_IAR/src/Blinky_Demo/main_blinky.c
src/FreeRTOS_Demo/Full_Demo/IntQueueTimer.c : derived from RX700_RX71M_RSK_GCC_e2studio_IAR/src/Full_Demo/IntQueueTimer.c
src/FreeRTOS_Demo/Full_Demo/IntQueueTimer.h : derived from RX700_RX71M_RSK_GCC_e2studio_IAR/src/Full_Demo/IntQueueTimer.h
src/FreeRTOS_Demo/Full_Demo/main_full.c     : derived from RX700_RX71M_RSK_GCC_e2studio_IAR/src/Full_Demo/main_full.c including RegTest inline asm codes.
src/FreeRTOS_Demo/Full_Demo/seria.c         : derived from RX700_RX71M_RSK_GCC_e2studio_IAR/src/cg_src/r_cg_sci_user_gcc.c having big difference.
src/FreeRTOS_Demo/demo_main.h               : new file
src/FreeRTOS_Demo/demo_specific_io.h        : new file
src/frtos_config/FreeRTOSConfig.h           : generated file but definition of configUSE_TASK_DPFPU_SUPPORT is added.
src/frtos_startup/freertos_start.c          : generated file but demo_main() and prvSetupHardware() are added and generated hook functions are modified.
src/smc_gen/general/r_cg_hardware_setup.c   : generated file but LED port and serial initialization are added in R_FIT_Board_Support_Settings().
```

**Comment by n9wxu** [comment]
Needs a rebase before the final approval can happen.

File: FreeRTOS/Demo/RX700_RX72N_EnvisionKit_GCC_e2studio/src/frtos_startup/freertos_start.c
Comment (n9wxu) [inline FreeRTOS/Demo/RX700_RX72N_EnvisionKit_GCC_e2studio/src/frtos_startup/freertos_start.c:78]: the pcTaskName needs to be a char* to match the definition in task.c.
That extern definition in task.c will shortly move to a prototype in task.h so this will generate a warning as the prototypes will not match.

Comment (NoMaY-jp) [reply FreeRTOS/Demo/RX700_RX72N_EnvisionKit_GCC_e2studio/src/frtos_startup/freertos_start.c:78]: Thank you for your checking.

I will merge recent FreeRTOS-kernel's (and FreeRTOS's) commits and fix the warning.

**Comment by NoMaY-jp** [comment]
I'm sorry that I may mistake something to do. I will check what I did...

**Comment by n9wxu** [comment]
No problem, I suspect the signed char issue has been around for a while and not limited to this port.  We discovered the issue while refactoring task.[c,h].

**Comment by NoMaY-jp** [comment]
I fixed the possiblity of warning which will be caused in the future by the commit a4c4177. Other commits are merges of recent FreeRTOS-kernel's and FreeRTOS's commits by other peoples. If I mistake operation of Git/GitHub, please let me know. (This is my first pull request.)

By the way, I notice that this incorrect argument type is due to the generated source file by e2 studio. So, I will report this issue to Renesas Electronics Corporation.

**Comment by NoMaY-jp** [comment]
Hello @n9wxu, (from https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/121#issuecomment-675021048.)

> We need register context switch test support

I think that there may be the following two points.

(1) Missing RegTest.s.
(2) Missing register context switch test of DPFPU's registers.

Regarding to the (1), I combined it into main_full.c as Richard did so in newer demo.

Richard's Demo/RX700_RX71M_RSK_Renesas_e2studio/src/Full_Demo/main_full.c:
https://github.com/FreeRTOS/FreeRTOS/tree/V10.3.1/FreeRTOS/Demo/RX700_RX71M_RSK_Renesas_e2studio/src/Full_Demo/main_full.c#L479

My Demo/{RX700_RX72N_EnvisionKit_Renesas_e2studio_CS%2B, and others}/src/FreeRTOS_Demo/Full_Demo/main_full.c:
https://github.com/FreeRTOS/FreeRTOS/blob/59f1d3e2fe4318adf1af0e79fae69469c0b3ad60/FreeRTOS/Demo/RX700_RX72N_EnvisionKit_Renesas_e2studio_CS%2B/src/FreeRTOS_Demo/Full_Demo/main_full.c#L480
https://github.com/FreeRTOS/FreeRTOS/blob/59f1d3e2fe4318adf1af0e79fae69469c0b3ad60/FreeRTOS/Demo/RX700_RX72N_EnvisionKit_GCC_e2studio/src/FreeRTOS_Demo/Full_Demo/main_full.c#L480
https://github.com/FreeRTOS/FreeRTOS/blob/59f1d3e2fe4318adf1af0e79fae69469c0b3ad60/FreeRTOS/Demo/RX700_RX72N_EnvisionKit_IAR_e2studio_EWRX/src/FreeRTOS_Demo/Full_Demo/main_full.c#L480

(Q) Do you think that it is better for me to separate out this part from main_full.c to RegTest.c?

Regarding to the (2), I understand that I forgot to add context switch test of DPFPU's registers. I'd like to add such test in a week (if no big trouble). But ...

(Q) Will my addition disturb your work?

Best Regards,
NoMaY

**Comment by NoMaY-jp** [comment]
Hello @n9wxu,

> > We need register context switch test support
> 
> I think that there may be the following two points.
> 
> (1) Missing RegTest.s.
> (2) Missing register context switch test of DPFPU's registers.

Is there no problem for me to do the following?

> Regarding to the (1), I combined it into main_full.c as Richard did so in newer demo.
> 
> Richard's Demo/RX700_RX71M_RSK_Renesas_e2studio/src/Full_Demo/main_full.c:
> https://github.com/FreeRTOS/FreeRTOS/tree/V10.3.1/FreeRTOS/Demo/RX700_RX71M_RSK_Renesas_e2studio/src/Full_Demo/main_full.c#L479
> ... omit ...
> 
> (Q) Do you think that it is better for me to separate out this part from main_full.c to RegTest.c?

I will keep current combined source code.

> Regarding to the (2), I understand that I forgot to add context switch test of DPFPU's registers. I'd like to add such test in a week (if no big trouble). But ...
> 
> (Q) Will my addition disturb your work?
> 

I will do it.

Best Regards,
NoMaY

**Comment by NoMaY-jp** [comment]
Hello n9wxu,

> We need register context switch test support

I fixed it by updating main_full.c at the commit e636eb8. (There are three main_full.c but these are the same source files.)

Best Regards,
NoMaY

**Review by n9wxu — APPROVED** [review]

**Comment by mingyue86010** [comment]
Hi @NoMaY-jp ,

There is a confilct in this PR. It looks the FreeRTOS/Source was somehow removed in your repo. Can you fix that so we can have this PR merged.

Thank you!

Regards,
Ming Yue

**Comment by NoMaY-jp** [comment]
I'm sorry for my mistake. I'm trying to fix conflict.

**Comment by NoMaY-jp** [comment]
I'm sorry for my mistake again. I'm now trying to fix unnecessary update of FreeRTOS/Source and SiFive_HiFive1_RTOS_demo.url.

**Comment by NoMaY-jp** [comment]
Hello @mingyue86010 and @n9wxu,

Thank you very much for taking care of this PR.

I have finished to fix conflict and other mistakes.

Best regards,
NoMaY

**Review by mingyue86010 — APPROVED** [review]

**Comment by mingyue86010** [comment]
> Hello @mingyue86010 and @n9wxu,
> 
> Thank you very much for taking care of this PR.
> 
> I have finished to fix conflict and other mistakes.
> 
> Best regards,
> NoMaY

@NoMaY-jp Thank you for your quick fix!

**Review by n9wxu — APPROVED** [review]

**Pull request merged**
